File: 2020_10/NumberOfRecentCalls/number_of_recent_calls.py
from collections import deque


class RecentCounter:

    def __init__(self):
        """
            I use list at first time I solve this question, but it need to l.pop(0) 
            and it need to move whole elements in list to their left side, it will casue some wasting.
            using deque make left pop more efficiently
        """
        self.ping_store = deque()

    def ping(self, t: int) -> int:
        lower = t - 3000
        self.ping_store.append(t)
        while lower > self.ping_store[0]:
            self.ping_store.popleft()
        return len(self.ping_store)


# Your RecentCounter object will be instantiated and called as such:
# obj = RecentCounter()
# param_1 = obj.ping(t)

File: 2020_10/NumberOfRecentCalls/numberOfRecentCalls.go
type RecentCounter struct {
    pingStore []int
}


func Constructor() RecentCounter {
    return RecentCounter{}
}


func (this *RecentCounter) Ping(t int) int {
    this.pingStore = append(this.pingStore, t)
    lower := t - 3000
    for lower > this.pingStore[0] {
        this.pingStore = this.pingStore[1:]
    }
    return len(this.pingStore)
}


/**
 * Your RecentCounter object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Ping(t);
 */